packages/core/src/services/loopDetectionService.ts

```diff
@@ -0,0 +1,85 @@
+/**
+ * @license
+ * Copyright 2025 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { createHash } from 'crypto';
+import { GeminiEventType, ServerGeminiStreamEvent } from '../core/turn.js';
+
+const TOOL_CALL_LOOP_THRESHOLD = 5;
+const CONTENT_LOOP_THRESHOLD = 10;
+const MAX_LOOPBACK_WINDOW = 1000;
+
+export class LoopDetectionService {
+  private toolCallCounts = new Map<string, number>();
+  private accumulatedContent = '';
+
+  private getToolCallKey(toolCall: { name: string; args: object }): string {
+    // Stringify args for a consistent key.
+    const argsString = JSON.stringify(toolCall.args);
+    const keyString = `${toolCall.name}:${argsString}`;
+    return createHash('sha256').update(keyString).digest('hex');
+  }
+
+  addAndCheck(event: ServerGeminiStreamEvent): boolean {
```

Comment:
We'll want to be conscious of how much this allocates. We're doing a lot of heavy string manipulation here. No immediate ask BUT something to monitor on release.